tests: Update more tests to work with descriptor wallets

achow101 opened the pull request:
I went through all the tests and checked whether they passed with descriptor wallets. This partially informed some changes in #16528. Some tests needed changes to work with descriptor wallets. These were primarily due to import and watchonly behavior. There are some tests and test cases that only test legacy wallet behavior so those tests won't be run with descriptor wallets.

This PR updates more tests to have to the `--descriptors` switch in `test_runner.py`. Additionally a mutually exclusive `--legacy-wallet` option has been added to force legacy wallets. This does nothing currently but will be useful in the future when descriptor wallets are the default. For the tests that rely on legacy wallet behavior, this option is being set so that we don't forget in the future. Those tests are `feature_segwit.py`, `wallet_watchonly.py`, `wallet_implicitsegwit.py`, `wallet_import_with_label.py`, and `wallet_import_with_label.py`.

If you invert the `--descriptors`/`--legacy-wallet` default so that descriptor wallets are the default, all tests (besides the legacy wallet specific ones) will pass.

achow101 committed:
Update wallet_importprunedfunds to avoid dumpprivkey

Removes the use of dumpprivkey so that descriptor wallets can pass on
this. Also does a few descriptor wallet specific changes due to
different IsMine semantics.

achow101 committed:
Use a separate watchonly wallet in rpc_fundrawtransaction.py

Import things into a separate watchonly wallet to work with descriptor
wallets.

achow101 committed:
Make import tests in wallet_listtransactions.py legacy wallet only

Existing import* RPCs are disabled for descriptor wallets, so only do
these tests for legacy wallets.

achow101 committed:
Avoid dumpprivkey in wallet_listsinceblock.py

Generate a privkey in the test framework instead of using dumpprivkey so
that descriptor wallets work in this test.

achow101 committed:
Use importdescriptors for descriptor wallets in wallet_bumpfee.py

If using descriptor wallets, use importdescriptors instead of
importmulti.

achow101 committed:
Move import and watchonly tests to be legacy wallet only in wallet_balance.py

Imports and watchonly behavior are legacy wallet only, so make them only
run when the test is in legacy wallet mode.

achow101 committed:
Use separate watchonly wallet for multisig in feature_nulldummy.py

Create and import the multisig into a separate watchonly wallet so that
feature_nulldummy.py works with descriptor wallets.

blocktools.create_raw_transaction is also updated to use multiple nodes
and wallets and to use PSBT so that this test passes.

achow101 committed:
Add descriptor wallet output to tool_wallet.py

Descriptor wallets output slightly different information in the wallet
tool, so check that output when in descriptor wallet mode.

achow101 committed:
Add script equivalent of functions in address.py

achow101 committed:
Avoid dumpprivkey and watchonly behavior in rpc_signrawtransaction.py

dumpprivkey and watchonly behavior don't work with descriptor wallets.

Test for multisigs is modified to not rely on watchonly behavior for
those multisigs. This has a side effect of removing listunspent, but
that's not the target of this test, so that's fine.

achow101 committed:
Use importdescriptors when in descriptor wallet mode in wallet_createwallet.py

sethdseed and importmulti are not available for descriptor wallets, so
when doing descriptor wallet tests, use importdescriptors instead.

Also changes some output to match what descriptor wallets will return.

achow101 committed:
Do addmultisigaddress tests in legacy wallet mode in wallet_address_types.py

addmultisigaddress is not available in descriptor wallets, so only run
these when testing legacy wallets

achow101 committed:
Make raw multisig tests legacy wallet only in rpc_rawtransaction.py

The traditional multisig workflow doesn't work with descriptor wallets
so make these tests legacy wallet only.

achow101 committed:
tests: Add a --legacy-wallet that is mutually exclusive with --descriptors

Although legacy wallet is still the default, for future use, add a
--legacy-wallet option to the test framework. Additional tests for
descriptor wallets have been enabled with the --descriptors option.
Tests that must be legacy wallet only are being started with
--legacy-wallet. Even though this option does not currently do anything,
this will be helpful in the future when descriptor wallets become the
default.

MarcoFalke commented:
Concept ACK, but I'd prefer if we got the current tests green first: #18800

Sjors commented:
Concept ACK. Needs rebase, now that #18805 landed, which fixed #18800.

Sjors commented:
Tests pass on macOS. The commit messages could be more descriptive than "Fix wallet_importprunedfunds", if only to aid the lazy reviewer in understanding what problems you're solving.

MarcoFalke commented:
Concept ACK

achow101 commented:
I've updated the commit messages to be more descriptive.

Sjors commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -480,22 +490,39 @@ def test_spend_2of2(self):
         oldBalance = self.nodes[1].getbalance()
         inputs = []
         outputs = {self.nodes[1].getnewaddress():1.1}
-        rawtx = self.nodes[2].createrawtransaction(inputs, outputs)
```
In e3dc50344a57b926373bcc51aa2e213a682051be it seems problematic to replace  `createrawtransaction` and `fundrawtransaction` with PSBT methods in `rpc_fundrawtransaction.py`.

achow101 replied:
I think it's fine in some instances (like this one) because the funding code is the same and the goal is to check that the funding works.

Sjors replied:
Do we still sufficiently test the `createrawtransaction` and `fundrawtransaction` RPC elsewhere? Those are rather ancient, with all cool kids using PSBT it might be a while before we see a regression.

achow101 replied:
IMO we do. `createrawtransaction` and `fundrawtransaction` are still being tested in `rpc_fundrawtransaction.py` as well is in other tests. Additionally, `fundrawtransaction` basically only calls `FundTransaction` after a bit of options parsing, and `walletcreatefundedpsbt` calls this function as well, right after transaction creation.

Sjors reviewed (commented):
f57cacf246375ab5f4338a2f169e72bc1cb3c63f looks good, except one issue, see inline.

0472e9f12c4c2efa5a0afaeee55f4311ffbd16f7: the commit message is confusing. Why is this code being ~duplicated?

achow101 commented:
> 0472e9f: the commit message is confusing. Why is this code being ~duplicated?

I need to get the scriptPubKeys for those things, not just the address. So the functions are partially duplicated to generate the scripts rather than the addresses.

Sjors commented:
Is there an easy way to convert scripts to addresses in the Python framework?

Anyway, ACK f57cacf

Sjors commented:
re-utACK 02357c7cd965ae5701d129691d4487a89f5d6deb, just some rebase related import changes

Sjors commented:
Possibly spurious Travis [error](https://travis-ci.org/github/bitcoin/bitcoin/jobs/722952265#L2561):
```
020-09-01T05:18:47.709000Z TestFramework (INFO): Initializing test directory /Users/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20200901_051608/wallet_basic_178
...
25662020-09-01T05:19:50.052000Z TestFramework (INFO): check -reindex
25672020-09-01T05:20:03.516000Z TestFramework (ERROR): Assertion failed
2568Traceback (most recent call last):
...
2575test_framework.authproxy.JSONRPCException: Amount out of range (-3)
2576
2577During handling of the above exception, another exception occurred:
2578
2579Traceback (most recent call last):
2580  File "/Users/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin16/test/functional/test_framework/test_framework.py", line 118, in main
2581    self.run_test()
2582  File "/Users/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin16/test/functional/wallet_basic.py", line 601, in run_test
2583    assert_raises_rpc_error(-6, "Transaction has too long of a mempool chain", self.nodes[0].sendtoaddress, sending_addr, node0_balance - Decimal('0.01'))
```

fjahr commented:
Concept ACK, will take a closer look soon.

Sjors commented:
re-utACK c45302f6f1d9e7e210b51f7bd580e498a60b2442

achow101 committed:
Update wallet_labels.py to not require descriptors=False

ryanofsky commented on test/functional/rpc_signrawtransaction.py:
```diff
@@ -151,21 +154,24 @@ def script_verification_error_test(self):
         self.nodes[0].generate(1)
         self.sync_all()
-        # Find the UTXO for the transaction node[1] should have received, check witnessScript matches
-        unspent_output = self.nodes[1].listunspent(0, 999999, [p2sh_p2wsh_address["address"]])[0]
```
In commit "Avoid dumpprivkey and watchonly behavior in rpc_signrawtransaction.py" (f4f81c98dca6da030b84b5b4887cf3ef4b72baec)

Can commit description be updated to mention this change and reason listunspent is not longer used or checked? Change seems reasonable but I don't get it currently

achow101 replied:
Updated the commit message. `listunspent` is not being used and is not the test target for this test. It is tested elsewhere. There is no need to attempt to bring this back for descriptor wallets.

ryanofsky commented on test/functional/test_framework/script_util.py:
```diff
@@ -24,3 +25,59 @@
+def check_script(script):
+    if (type(script) is str):
+        script = hex_str_to_bytes(script) # Assuming this is hex string
+    if (type(script) is bytes or type(script) is CScript):
```
In commit "Add script equivalent of functions in address.py" (a56fc6c8c8f58f4a6ade3ac1437875bf8071a844)

Not important, but it's usually better to say `isinstance(obj, obj_type)` instead of `type(obj) is obj_type` to not interfere with subclassing. Applies various places this commit

achow101 replied:
Done

ryanofsky commented on test/functional/feature_nulldummy.py:
```diff
@@ -51,10 +51,17 @@ def skip_test_if_missing_module(self):
+        self.ms_address = wmulti.addmultisigaddress(1, [self.pubkey])['address']
+        self.wit_address = w0.getnewaddress(address_type='p2sh-segwit')
+        self.wit_ms_address = wmulti.addmultisigaddress(1, [self.pubkey], '', 'p2sh-segwit')['address']
+        if not self.options.descriptors:
```
In commit "Use separate watchonly wallet for multisig in feature_nulldummy.py" (ed97c9b67e715815f653e20f5e478a4dad4e13f8)

Can a comment be added to explain why the non-descriptor version of the test needs these addresses imported to pass, while the descriptor version doesn't need them? (Assuming that is the case.) I'm missing some understanding here

achow101 replied:
Legacy wallets need these imported for the test to work. Otherwise they won't watch the multisig. Descriptor wallets will always watch imported things.

achow101 replied:
Added a comment.

ryanofsky commented on test/functional/wallet_balance.py:
```diff
@@ -57,14 +57,15 @@ def skip_test_if_missing_module(self):
-        # Check that nodes don't own any UTXOs
-        assert_equal(len(self.nodes[0].listunspent()), 0)
-        assert_equal(len(self.nodes[1].listunspent()), 0)
+        if not self.options.descriptors:
```
In commit "Move import and watchonly tests to be legacy wallet only in wallet_balance.py" (8ee94451ad55b81a53c38c6b8cee0c13f1b1f9e7)

It would be helpful to add the comment here from the commit description: "Imports and watchonly behavior are legacy wallet only, so make them only run when the test is in legacy wallet mode." It would also be helpful to say whether it is a good thing to be skipping these checks, or if the test maybe should be port these checks to work with a watchonly descriptor wallet in the future

achow101 replied:
Added a comment. It is not necessary to test these for descriptor wallets, that's why they are disabled in the first place. Otherwise there would be a TODO.

ryanofsky commented on test/functional/wallet_balance.py:
```diff
@@ -156,7 +162,12 @@ def test_balances(*, fee_node_1=0):
                                                  'trusted':           Decimal('0E-8'),  # node 1's send had an unsafe input
                                                  'untrusted_pending': Decimal('30.0') - fee_node_1}}  # Doesn't include output of node 0's send since it was spent
-            assert_equal(self.nodes[0].getbalances(), expected_balances_0)
+            if self.options.descriptors:
```
In commit "Move import and watchonly tests to be legacy wallet only in wallet_balance.py" (8ee94451ad55b81a53c38c6b8cee0c13f1b1f9e7)

Would suggest writing this as:

```python
if self.options.descriptors:
   del expected_balances_0['watchonly']
assert_equal(self.nodes[0].getbalances(), expected_balances_0)
```

to keep checking for new fields as strict and to duplicate less code

achow101 replied:
Done

ryanofsky commented on test/functional/wallet_balance.py:
```diff
@@ -246,7 +257,7 @@ def test_balances(*, fee_node_1=0):
         self.nodes[1].sendrawtransaction(hexstring=tx_replace, maxfeerate=0)
 
         # Now confirm tx_replace
-        block_reorg = self.nodes[1].generatetoaddress(1, ADDRESS_WATCHONLY)[0]
```
In commit "Move import and watchonly tests to be legacy wallet only in wallet_balance.py" (8ee94451ad55b81a53c38c6b8cee0c13f1b1f9e7)

I don't get these changes. Does the address generated to matter for the tests or is just code cleanup? It would be good to mention these changes in the commit description and say what they are

achow101 replied:
Reverted.

ryanofsky reviewed (approved):
Code review ACK 2cc253dc7f385f774fa057eed5e2340ffbd022f2. All looks very good. My review comments below are just asking for comments & clarifications and aren't important

achow101 commented on test/functional/wallet_importprunedfunds.py:
```diff
@@ -30,8 +33,11 @@ def run_test(self):
         address2 = self.nodes[0].getnewaddress()
         # privkey
-        address3 = self.nodes[0].getnewaddress()
-        address3_privkey = self.nodes[0].dumpprivkey(address3)  # Using privkey
```
Not currently.

achow101 commented on test/functional/wallet_listtransactions.py:
```diff
@@ -91,18 +91,19 @@ def run_test(self):
-        assert_array_result(self.nodes[0].listtransactions(label="watchonly", count=100, include_watchonly=True),
-                            {"category": "receive", "amount": Decimal("0.1")},
-                            {"txid": txid, "label": "watchonly"})
+        if not self.options.descriptors:
```
It's not necessary because it is testing the `include_watchonly` parameter in `listtransactions` which is not present in descriptor wallets.

achow101 replied:
Added a comment. It is not necessary to test these for descriptor wallets, that's why they are disabled in the first place. Otherwise there would be a TODO.

achow101 committed:
Disable some tests for tool_wallet when descriptors

Some tests are legacy wallet only (and make legacy wallets) so they
shouldn't be run when doing descriptor wallet tests.

achow101 committed:
Update feature_backwards_compatibility for descriptor wallets

achow101 commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -242,121 +243,122 @@ def run_test(self):
-        self.nodes[0].generate(1)
-        self.sync_all()
-        assert_equal(self.nodes[0].getbalance(), bal+Decimal('50.00000000')+Decimal('2.19000000')) #block reward + tx
+        if not self.options.descriptors:
```
Added a comment. It is not necessary to test these for descriptor wallets, that's why they are disabled in the first place. Otherwise there would be a TODO.

achow101 commented on test/functional/wallet_address_types.py:
```diff
@@ -228,18 +228,23 @@ def run_test(self):
+            self.test_address(3, self.nodes[3].addmultisigaddress(2, [compressed_1, compressed_2])['address'], True, 'bech32')
+
+        do_multisigs = [False]
+        if not self.options.descriptors:
```
Added a comment. It is not necessary to test these for descriptor wallets, that's why they are disabled in the first place. Otherwise there would be a TODO.

achow101 commented:
Going to also add more changes to `wallet_multiwallet.py` soon.

achow101 commented:
Actually I'll deal with multiwallet separately. It's going to be involved in a followup to drop the wallet dirs thing.

ryanofsky reviewed (approved):
Code review ACK bf7142f0f0ce5b8d8a65b9732c7ccaccbe718f94. Changes since last review: two new commits, lots of comments and extra commit message information in existing commits, and test runner change to avoid running some legacy tests twice

achow101 committed:
Avoid creating legacy wallets in wallet_importdescriptors.py

achow101 committed:
tests: Make only desc wallets for wallet_multwallet.py --descriptors

achow101 commented:
Rebased after #20186 and also fixed `wallet_multiwallet.py`.

ryanofsky reviewed (approved):
Code review ACK 5d24a64ab8aff27f0df567bd8421f07dbe8c5b2f. Changes since last review: rebase, isinstance fixes, compatibility test fixes, and two new commits adding more --descriptors support. New commit using descriptors more widely in wallet_multiwallet.py makes the test coverage a lot better, and the code clearer too

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -215,67 +216,89 @@ def run_test(self):
+            assert info['private_keys_enabled'] == False
+            assert info['keypoolsize'] == 0
+        else:
+            # Descriptor wallets appear to corrupted wallets to old software
```
nit to->as?

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -284,39 +307,35 @@ def run_test(self):
         node_v17.assert_start_raises_init_error(["-wallet=w3_v18"], "Error: Error loading w3_v18: Wallet requires newer version of Bitcoin Core")
-        node_v17.assert_start_raises_init_error(["-wallet=w3"], "Error: Error loading w3: Wallet requires newer version of Bitcoin Core")
+        if self.options.descriptors:
+            # Descriptor wallets appear to corrupted wallets to old software
```
nit: to->as

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -261,9 +261,12 @@ def run_test(self):
-        self.test_getwalletinfo_on_different_wallet()
-        self.test_salvage()
+        if not self.options.descriptors:
+            # TODO: Wallet tool needs more create options at which point these can be enabled.
```
Running the tests on commit 80bb5ca8cf fails.

This commit might need to come first?

maflcko reviewed (approved):
quick cr ACK 5d24a64ab8

MarcoFalke commented on test/functional/wallet_labels.py:
```diff
@@ -156,7 +156,7 @@ def run_test(self):
+                "Address is not valid" if self.options.descriptors else "Invalid Bitcoin address or script",
                 lambda: wallet_watch_only.importaddress(label=l, rescan=False, address=ad),
             )
 
```
in commit c1f7bfc6bb1074c26d670976a8f8c0bb95be1858:

Should wallet_labels be run with and without descriptors in the test_runner?

mjdietzx commented on test/functional/wallet_createwallet.py:
```diff
@@ -113,11 +147,15 @@ def run_test(self):
         walletinfo = w6.getwalletinfo()
-        assert_equal(walletinfo['keypoolsize'], 1)
-        assert_equal(walletinfo['keypoolsize_hd_internal'], 1)
+        if self.options.descriptors:
+            assert_equal(walletinfo['keypoolsize'], 3)
+            assert_equal(walletinfo['keypoolsize_hd_internal'], 3)
+        else:
+            assert_equal(walletinfo['keypoolsize'], 1)
+            assert_equal(walletinfo['keypoolsize_hd_internal'], 1)
```
Would this be a little easier to follow? Also, is the comment on l148 still relevant `# There should only be 1 key`?
```python
nkeys = 3 if self.options.descriptors else 1
assert_equal(walletinfo['keypoolsize'], nkeys)
assert_equal(walletinfo['keypoolsize_hd_internal'], nkeys)
```
```

achow101 replied:
Done

MarcoFalke replied:
I thin the previous version of the code was more correct. See ci failure

achow101 replied:
Oops, I forgot to change the second assert. Should be fixed now.

mjdietzx commented on test/functional/test_framework/blocktools.py:
```diff
@@ -173,14 +172,18 @@ def create_transaction(node, txid, to_address, *, amount):
-    signresult = node.signrawtransactionwithwallet(rawtx)
-    assert_equal(signresult["complete"], True)
-    return signresult['hex']
+    psbt = node.createpsbt(inputs=[{"txid": txid, "vout": 0}], outputs={to_address: amount})
+    for _ in range(2):
+        for w in node.listwallets():
+            wrpc = node.get_wallet_rpc(w)
+            signed_psbt = wrpc.walletprocesspsbt(psbt)
+            psbt = signed_psbt['psbt']
+    final_psbt = node.finalizepsbt(psbt)
+    assert_equal(final_psbt["complete"], True)
+    return final_psbt['hex']
```
I haven't been able to wrap my head around this / how this change fits into the greater PR. If you don't mind could you give me a brief explanation or point me in the right direction?

achow101 replied:
This change makes this function use the PSBT workflow rather than the old raw transactions workflow. Additionally, it tries to sign with all of the wallets loaded by the node.

ryanofsky replied:
> I haven't been able to wrap my head around this / how this change fits into the greater PR. If you don't mind could you give me a brief explanation or point me in the right direction?

My understanding is that legacy wallets support a mix of watch only and spendable scriptpubkeys, while descriptor wallets have to choose between only watch-only or spendable descriptors and refuse to hold both. So while a single wallet worked in many of these tests previously, now two wallets are required to keep all of the same test coverage. In some cases like this the PR creates an extra watchonly wallet to maintain coverage. In other cases it keeps the coverage for legacy wallets but drops it for descriptors to avoid needing the extra wallet.

I think the descriptor watchonly limitation is supposed to be somewhat artificial currently, but in the future if every wallet is required to be fully watchonly or not, then GUI & RPC interfaces may be simplified later.

MarcoFalke commented:
Almost cr ACK d137ec7d8499497fde60e23d0cb5fc7f7387e5e3 🍱

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Almost cr ACK d137ec7d8499497fde60e23d0cb5fc7f7387e5e3 🍱
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj9cAwAvAxznEyfwZnurWI8ACsC3a6Y0Hj28zdqM99HbnkyRdrwcnXhLpgr++pS
9sJw03kdVgx6XqYOhet6naCU3AieNexe92TcLaFzTIVxnWVPAsPr97og6w3CpP65
SM6HHQdo2bUC1CAum/hq9oj0n9oQ8EAFYBAJhNNfArIEvGKMsO2eai8QlggQU23B
aiIo+pi+do3GAJ02zjXH9ed7kQJVsobOhO/KCBWQVGqQQjj3sC4t++A9kE41LKu/
Jv0U7pxpEF7J3ghy9fLPeDLP1QadIsZIQTa/kBuEzLMGdrmr1iTAmPT46vd6VORd
fUevN58HqkUy+cihFjjT11YxASbeVy9eFrI5ldswZIePN5PUyGj6NVdOCc9LTz8+
CX1RNw/V7cVSwgk05GA/zQVSRRxjkTDJHpS23i1cxliUiQ17ryQeVimBfWnl0CCt
6fuZa7iMzDt0AyAP231bSuiIuQwG1x9CsaYVZZA9mFa8C7IzaUlR4zO8dwm5s9EG
B3zeS/DY
=XRP+
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `64884ac60aea3836fd1197bfc420b7916e0c7bc4009dbc1c6a7154bab7e960d9  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010864884ac60aea3836fd1197bfc420b7916e0c7bc4009dbc1c6a7154bab7e960d9f010835bcda451bcbc4f328469bfcf87556608fff0102c5ccde6f4ea4632d41afccd49a1e56208f1045f9d1d46f008a657f0330c55a8530083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0102ec5f1368ede150f71e05f94601b46a108f020d268737ec93ed3457d4aacc1524d0b413b251aeb6b399783bf628da3b02d858c08f1045f9d1d49f008b3603d0fa8d384ab0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010302c4a58740719a9e28776251a9fa3ed08f12014ccc053b088915aee2218e78cc8cdd465452c1d46e069b013fdaf9c640984a208f1045f9d1d46f008aa9441cd2cbc75ef0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0108ad62581d833e1eb65e7c4b3da45e16c08f1045f9d1d47f008527ffe37b36fdf320083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
review ACK c7b7e0a69265946aecc885be911c7650911ba2e3 🎿

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK c7b7e0a69265946aecc885be911c7650911ba2e3 🎿
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUixwgv+P+5WXXpHmaWiW4IZOO+ZwjR3ywM62Ifs63hMf83bFO0cwDH3NkwzY7hn
+qNiNx7qexza6nJUsiy0EcekBo743AGYx2Ttun33F+Dj9yN9I/aGZPJ3sIIc5grU
LXWfYgwgeQcb1Kkovkv5viEOo0riA4SksdCySxQkybreWgeCV5hfiz+aS95WJx6w
90cGwfLqoj0ZOsAhE59Y4cNK0ZGZ/8MuJ+DLTT+uyJBnHAaxrGBFPtWx1MG+Npzq
AKBHPO8d554nAp5Vw3blf/ZjKB++mCB0iN6gR0Ja3Mtne1YetnYXEVKHQ/Yfvo2F
po/GIccCibNb1TvFjtnWUeX12bkBCqzXIKg0JAP4V6+THR6o3DwpeVIO4drk/oVf
dxb1EVW/IugsM5diPhfyicbvYJavodB7eR8e03oVZv7JkBJmvyuuHLweKn+Ziuol
0ZuU6GxJGECuTT2r7VA4gwop/pueJ4l9lCm+RO9yTnfRLgIZ6gCq0lfQQctTlpWq
d7wv/9hm
=LD2k
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4756f18231367024c68d5255b8972d1d0c9ad9f91dd41296c63c8c42b777c68b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084756f18231367024c68d5255b8972d1d0c9ad9f91dd41296c63c8c42b777c68bf010885af14e7cb6c44c3444dfd8ac480e6c08fff010534d988f9022b16053b75883718522d408f1045f9fd8abf008a1176536e89de9560083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01087913d7e47c622a5a88005755fe094b208f1045f9fd8aaf0086799e56ca9e2aaa30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010e63f4589a7fb84ab2470f2002eeb698208f0200c272752ad66973c8901f8b74b1d6f1c25e9ba73ab364812b6a9596452ea80e908f1045f9fd8aef008383d081ea894d5390083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e82f2e487423e9a0326b9acbd6d2787c08f1045f9fd8adf008f603b039769ec4030083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

laanwj commented:
ACK c7b7e0a69265946aecc885be911c7650911ba2e3
